src/workerd/api/system-streams.c++

```diff
@@ -252,11 +252,99 @@ kj::Own<WritableStreamSink> newSystemStream(
   return kj::heap<EncodedAsyncOutputStream>(kj::mv(inner), encoding, context);
 }
 
+
+class WrappedAsyncIoStream final :
+    public ReadableStreamSource, public WritableStreamSink, public kj::Refcounted {
+  // A wrapper around a native `kj::AsyncIoStream` to enable a ReadableStream and WritableStream
+  // to be constructed from it. This stream does not support encoding.
+
+public:
+  explicit WrappedAsyncIoStream(kj::Own<kj::AsyncIoStream> inner, IoContext& context);
+  ~WrappedAsyncIoStream();
+
+  // ReadableStreamSource methods:
+  kj::Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override;
+
+  kj::Maybe<uint64_t> tryGetLength(StreamEncoding outEncoding) override;
+
+  kj::Maybe<Tee> tryTee(uint64_t limit) override;
+
+  // WritableStreamSink methods:
+  kj::Promise<void> write(const void* buffer, size_t size) override;
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const kj::byte>> pieces) override;
+
+  kj::Maybe<kj::Promise<DeferredProxy<void>>> tryPumpFrom(
+      ReadableStreamSource& input, bool end) override;
+
+  kj::Promise<void> end() override;
+
+  void abort(kj::Exception reason) override;
+
+private:
+  kj::Own<kj::AsyncIoStream> inner;
+
+  IoContext& ioContext;
+};
+
+WrappedAsyncIoStream::WrappedAsyncIoStream(
+    kj::Own<kj::AsyncIoStream> inner, IoContext& context)
+    : inner(kj::mv(inner)), ioContext(context) { }
+
+WrappedAsyncIoStream::~WrappedAsyncIoStream() {
+  inner->shutdownWrite();
+}
+
+kj::Promise<size_t> WrappedAsyncIoStream::tryRead(
+    void* buffer, size_t minBytes, size_t maxBytes) {
+  return inner->tryRead(buffer, minBytes, maxBytes)
+      .attach(ioContext.registerPendingEvent());
+}
+
+kj::Maybe<uint64_t> WrappedAsyncIoStream::tryGetLength(StreamEncoding outEncoding) {
+  KJ_REQUIRE(outEncoding == StreamEncoding::IDENTITY);
+  return inner->tryGetLength();
+}
+
+kj::Maybe<ReadableStreamSource::Tee> WrappedAsyncIoStream::tryTee(uint64_t limit) {
+  auto tee = kj::newTee(kj::mv(inner), limit);
+
+  Tee result;
+  result.branches[0] = newSystemStream(
+      newTeeErrorAdapter(kj::mv(tee.branches[0])), StreamEncoding::IDENTITY);
+  result.branches[1] = newSystemStream(
+      newTeeErrorAdapter(kj::mv(tee.branches[1])), StreamEncoding::IDENTITY);
+  return kj::mv(result);
+}
+
+kj::Promise<void> WrappedAsyncIoStream::write(const void* buffer, size_t size) {
+  return inner->write(buffer, size).attach(ioContext.registerPendingEvent());
+}
+
+kj::Promise<void> WrappedAsyncIoStream::write(
+    kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) {
+  return inner->write(pieces).attach(ioContext.registerPendingEvent());
+}
+
+kj::Maybe<kj::Promise<DeferredProxy<void>>> WrappedAsyncIoStream::tryPumpFrom(
+    ReadableStreamSource& input, bool end) {
+  return input.pumpTo(*this, end);
```

Comment:
I think this will recurse forver -- check out the definition of `ReadableStreamSource::pumpTo()`: https://github.com/cloudflare/workerd/blob/main/src/workerd/api/streams/internal.c%2B%2B#L333

Instead, this is our opportunity to optimize the pump. We can try downcasting `ReadableStreamSource` to a `WrappedAsyncIoStream` (see the first line of `EncodedAsyncOutputStream::tryPumpFrom()` for an example) and, if that returns non-null, this optimizes to `wrappedInput->inner->pumpTo(*inner)`, and we can even return a deferred proxy promise.

If `input` is _not_ a `WrappedAsyncIoStream`, then the correct thing to do is return nullptr, which causes the calling `pumpTo()` to fall back to its naive implementation.

If you just want to get things going, I think it'd be correct to unconditionally return nullptr, just suboptimal.

Reply:
Oh, this seems like a good argument not to introduce a new type at all, and instead try to reuse `EncodedAsync*Stream`. What if you are pumping from an `EncodedAsyncInputStream` to a `WrappedAsyncIoStream`? This pump should be optimized, too, as should `WrappedAsyncIoStream` -> `EncodedAsyncOutputStream`.

If we keep using the encoded streams then this all just works without quadratic checks. Setting the encoding to `IDENTITY` seems fine.

The only thing is, we need `EncodedAsyncOutputStream::end()` to call `shutdownWrite()`, which is awkward because it doesn't know that the inner stream is an `AsyncIoStream`.

Actually, is this buggy today? It's weird that `end()` does not indicate EOF to the underlying stream at all, unless the `EncodedAsyncOutputStream` itself is dropped immediately after...

Reply:
@dom96 This was marked resolved but it looks like while Harris's comment was addressed, mine was not.

Reply:
@kentonv Can you clarify which part?

I do reuse the `EncodedAsync*Stream` here: https://github.com/cloudflare/workerd/pull/237/files#diff-5f332a1e13274be21437d308032286aa3197c8158f8a9274c814296b3147691fR298 and I do call `shutdownWrite` here: https://github.com/cloudflare/workerd/pull/237/files#diff-5f332a1e13274be21437d308032286aa3197c8158f8a9274c814296b3147691fR336.

Reply:
@harrishancock appreciate you taking a look as well since we discussed how to solve this, would like to make sure I didn't miss anything in my implementation

Reply:
Oh I see, I missed the fact that `EncodedAsyncIoStream` is now a wrapper around a pair of `EncodedAsyncInputStream` and `EncodedAsyncOutputStream`.

But why? `newSystemMultiStream()` itself returns a pair of objects `readable` and `writable`, but as currently implemented they are both refcounts of a single `EncodedAsyncIoStream`, which in turn wraps a pair of `EncodedAsyncInputStream` and `EncodedAsyncOutputStream`.

Why not have `newSystemMultiStream()` just return the separate `EncodedAsyncInputStream` and `EncodedAsyncOutputStream`, and delete the `EncodedAsyncIoStream` class?

Reply:
IIRC I need the wrapper `EncodedAsyncIoStream` class to be able to call `shutdownWrite` on the `AsyncIoStream` when `end()` is called.

Reply:
I don't think you do. `EncodedAsyncOutputStream::end()` can call `shutdownWrite()`. A couple possible approaches:

* It could use `dynamic_cast` to test if the inner stream implements `AsyncIoStream` and, if so, call `shutdownWrite()`.
* It could have an additional constructor which accept `AsyncIoStream`, and it could store multiple references -- `kj::Own<kj::AsyncInputStream> inner` and `kj::Maybe<kj::AsyncIoStream&> ioInner` -- the latter is only non-null if the input was AsyncIoStream.

The latter approach is somewhat cleaner but also more involved. I would be fine with either.

Incidentally in the case that the inner object is _not_ `AsyncIoStream`, but just regular `AsyncInputStream`, then the way to signal EOF is to destroy the object. But, `end()` isn't do that today! This seems like a bug. I suppose it only works because the owner of the `WritableStreamSink` always drops the object immediately after `end()` finishes?

Reply:
hm, okay, I updated to use `dynamic_cast` though I think I am missing why this approach is significantly better than what I had. Is it mainly the reduction in indirection?

I didn't add any further destruction of the inner object if it's not an `AsyncIoStream`, if you think I should add that can you let me know what the best way to destroy the object explicitly there is?

Reply:
Well, I had been hoping you could get rid of the stream wrapper altogether, but I see you still have it just for the sake of refcounting. Maybe you could do the same trick we did in the http-over-capnp PR?

```suggestion
  struct RefcountedIoStream: public kj::Refcounted {
    kj::Own<AsyncIoStream> inner;
    RefcountedIoStream(kj::Own<AsyncIoStream> inner): inner(kj::mv(inner)) {}
  };
  auto rc = kj::refcounted<RefcountedIoStream>(kj::mv(stream));
  kj::Own<kj::AsyncIoStream> ref1 = kj::attachRef(*inner.get(), kj::addRef(*rc));
  kj::Own<kj::AsyncIoStream> ref2 = kj::attachRef(*inner.get(), kj::mv(*rc));
```

(Perhaps we should add this pattern to KJ...)

Reply:
I made a KJ PR to formalize this pattern: https://github.com/capnproto/capnproto/pull/1605

Reply:
I spent some time trying to figure out how to do this again and completely forgot that used this pattern in the http-over-capnp PR. Anyway, thanks for creating that PR, I'll wait until it gets merged and modify this PR to use it.

Reply:
Merged. Don't forget we'll need to update the capnproto submodule on the internal repo as well. (Something I myself forgot to do recently with a workerd PR...)